File: VB/ReadOnlyCollectionWrapper/ReadOnlyCollectionWrapper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections
Imports System.ComponentModel
Imports System.Collections.Generic

Namespace DXSample
	Friend Class ReadOnlyCollectionWrapper
		Inherits Component
		Implements IBindingList, ITypedList
		Private Shared ReadOnly fListChanged As Object = New Object()

		Public Sub New(ByVal dataSource As IBindingList)
			fDataSource = dataSource
		End Sub

		Public Sub New()
		End Sub

		Private fDataSource As IBindingList
		<AttributeProvider(GetType(IListSource))> _
		Public Property DataSource() As IBindingList
			Get
				Return fDataSource
			End Get
			Set(ByVal value As IBindingList)
				If fDataSource Is value Then
					Return
				End If
				If fDataSource IsNot Nothing Then
					RemoveHandler fDataSource.ListChanged, AddressOf OnDataSourceListChanged
				End If
				fDataSource = value
				If fDataSource IsNot Nothing Then
					AddHandler fDataSource.ListChanged, AddressOf OnDataSourceListChanged
				End If
				RaiseListChanged(New ListChangedEventArgs(ListChangedType.Reset, -1))
			End Set
		End Property

		Private Sub RaiseListChanged(ByVal args As ListChangedEventArgs)
			Dim handler As ListChangedEventHandler = TryCast(Events(fListChanged), ListChangedEventHandler)
			If handler IsNot Nothing Then
				handler(Me, args)
			End If
		End Sub

		Private Sub OnDataSourceListChanged(ByVal sender As Object, ByVal e As ListChangedEventArgs)
			RaiseListChanged(e)
		End Sub

		Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
			If disposing Then
				If fDataSource IsNot Nothing Then
					RemoveHandler fDataSource.ListChanged, AddressOf OnDataSourceListChanged
					fDataSource = Nothing
				End If
			End If
			MyBase.Dispose(disposing)
		End Sub

		#Region "IBindingList Members"

		Private Sub AddIndex(ByVal [property] As PropertyDescriptor) Implements IBindingList.AddIndex
			fDataSource.AddIndex([property])
		End Sub

		Private Function AddNew() As Object Implements IBindingList.AddNew
			Throw New NotSupportedException()
		End Function

		Private ReadOnly Property AllowEdit() As Boolean Implements IBindingList.AllowEdit
			Get
				Return True
			End Get
		End Property

		Private ReadOnly Property AllowNew() As Boolean Implements IBindingList.AllowNew
			Get
				Return False
			End Get
		End Property

		Private ReadOnly Property AllowRemove() As Boolean Implements IBindingList.AllowRemove
			Get
				Return False
			End Get
		End Property

		Private Sub ApplySort(ByVal [property] As PropertyDescriptor, ByVal direction As ListSortDirection) Implements IBindingList.ApplySort
			fDataSource.ApplySort([property], direction)
		End Sub

		Private Function Find(ByVal [property] As PropertyDescriptor, ByVal key As Object) As Integer Implements IBindingList.Find
			Return fDataSource.Find([property], key)
		End Function

		Private ReadOnly Property IsSorted() As Boolean Implements IBindingList.IsSorted
			Get
				Return fDataSource.IsSorted
			End Get
		End Property

		Private Custom Event ListChanged As ListChangedEventHandler Implements IBindingList.ListChanged
			AddHandler(ByVal value As ListChangedEventHandler)
				Events.AddHandler(fListChanged, value)
			End AddHandler
			RemoveHandler(ByVal value As ListChangedEventHandler)
				Events.RemoveHandler(fListChanged, value)
			End RemoveHandler
			RaiseEvent(ByVal sender As System.Object, ByVal e As System.ComponentModel.ListChangedEventArgs)
			End RaiseEvent
		End Event

		Private Sub RemoveIndex(ByVal [property] As PropertyDescriptor) Implements IBindingList.RemoveIndex
			fDataSource.RemoveIndex([property])
		End Sub

		Private Sub RemoveSort() Implements IBindingList.RemoveSort
			fDataSource.RemoveSort()
		End Sub

		Private ReadOnly Property IBindingList_SortDirection() As ListSortDirection Implements IBindingList.SortDirection
			Get
				Return fDataSource.SortDirection
			End Get
		End Property

		Private ReadOnly Property SortProperty() As PropertyDescriptor Implements IBindingList.SortProperty
			Get
				Return fDataSource.SortProperty
			End Get
		End Property

		Private ReadOnly Property SupportsChangeNotification() As Boolean Implements IBindingList.SupportsChangeNotification
			Get
				Return fDataSource.SupportsChangeNotification
			End Get
		End Property

		Private ReadOnly Property SupportsSearching() As Boolean Implements IBindingList.SupportsSearching
			Get
				Return fDataSource.SupportsSearching
			End Get
		End Property

		Private ReadOnly Property SupportsSorting() As Boolean Implements IBindingList.SupportsSorting
			Get
				Return fDataSource.SupportsSorting
			End Get
		End Property

		#End Region

		#Region "IList Members"

		Private Function Add(ByVal value As Object) As Integer Implements IList.Add
			Return fDataSource.Add(value)
		End Function

		Private Sub Clear() Implements IList.Clear
			fDataSource.Clear()
		End Sub

		Private Function Contains(ByVal value As Object) As Boolean Implements IList.Contains
			Return fDataSource.Contains(value)
		End Function

		Private Function IndexOf(ByVal value As Object) As Integer Implements IList.IndexOf
			Return fDataSource.IndexOf(value)
		End Function

		Private Sub Insert(ByVal index As Integer, ByVal value As Object) Implements IList.Insert
			fDataSource.Insert(index, value)
		End Sub

		Private ReadOnly Property IsFixedSize() As Boolean Implements IList.IsFixedSize
			Get
				Return fDataSource.IsFixedSize
			End Get
		End Property

		Private ReadOnly Property IsReadOnly() As Boolean Implements IList.IsReadOnly
			Get
				Return False
			End Get
		End Property

		Private Sub Remove(ByVal value As Object) Implements IList.Remove
			fDataSource.Remove(value)
		End Sub

		Private Sub RemoveAt(ByVal index As Integer) Implements IList.RemoveAt
			fDataSource.RemoveAt(index)
		End Sub

		Public Property IList_Item(ByVal index As Integer) As Object Implements IList.Item
			Get
				Return fDataSource(index)
			End Get
			Set(ByVal value As Object)
				fDataSource(index) = value
			End Set
		End Property

		#End Region

		#Region "ICollection Members"

		Private Sub CopyTo(ByVal array As Array, ByVal index As Integer) Implements ICollection.CopyTo
			fDataSource.CopyTo(array, index)
		End Sub

		Private ReadOnly Property Count() As Integer Implements ICollection.Count
			Get
				Return fDataSource.Count
			End Get
		End Property

		Private ReadOnly Property IsSynchronized() As Boolean Implements ICollection.IsSynchronized
			Get
				Return fDataSource.IsSynchronized
			End Get
		End Property

		Private ReadOnly Property SyncRoot() As Object Implements ICollection.SyncRoot
			Get
				Return fDataSource.SyncRoot
			End Get
		End Property

		#End Region

		#Region "IEnumerable Members"

		Private Function GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
			Return fDataSource.GetEnumerator()
		End Function

		#End Region

		#Region "ITypedList Members"

		Private Function GetItemProperties(ByVal listAccessors() As PropertyDescriptor) As PropertyDescriptorCollection Implements ITypedList.GetItemProperties
			Dim tList As ITypedList = TryCast(fDataSource, ITypedList)
			Dim properties As PropertyDescriptorCollection
			If tList Is Nothing Then
				If fDataSource.Count = 0 OrElse listAccessors IsNot Nothing Then
					properties = New PropertyDescriptorCollection(New PropertyDescriptor(){})
				Else
					properties = TypeDescriptor.GetProperties(fDataSource(0))
				End If
			Else
				properties = (CType(fDataSource, ITypedList)).GetItemProperties(listAccessors)
			End If
			Dim result As New List(Of PropertyDescriptor)()
			For Each [property] As PropertyDescriptor In properties
				result.Add(New ReadOnlyPropertyDescriptor([property]))
			Next [property]
			Return New PropertyDescriptorCollection(result.ToArray())
		End Function

		Private Function GetListName(ByVal listAccessors() As PropertyDescriptor) As String Implements ITypedList.GetListName
			Dim tList As ITypedList = TryCast(fDataSource, ITypedList)
			If tList Is Nothing Then
				If fDataSource.Count = 0 OrElse listAccessors IsNot Nothing Then
					Return String.Empty
				Else
					Return fDataSource(0).GetType().Name
				End If
			End If
			Return (CType(fDataSource, ITypedList)).GetListName(listAccessors)
		End Function

		#End Region
	End Class

	Public Class ReadOnlyPropertyDescriptor
		Inherits PropertyDescriptor
		Private fDescriptor As PropertyDescriptor

		Public Sub New(ByVal descr As PropertyDescriptor)
			MyBase.New(descr)
			fDescriptor = descr
		End Sub

		Public Overrides Function CanResetValue(ByVal component As Object) As Boolean
			Return False
		End Function

		Public Overrides ReadOnly Property ComponentType() As Type
			Get
				Return fDescriptor.ComponentType
			End Get
		End Property

		Public Overrides Function GetValue(ByVal component As Object) As Object
			Return fDescriptor.GetValue(component)
		End Function

		Public Overrides ReadOnly Property IsReadOnly() As Boolean
			Get
				Return True
			End Get
		End Property

		Public Overrides ReadOnly Property PropertyType() As Type
			Get
				Return fDescriptor.PropertyType
			End Get
		End Property

		Public Overrides Sub ResetValue(ByVal component As Object)
			Throw New NotSupportedException()
		End Sub

		Public Overrides Sub SetValue(ByVal component As Object, ByVal value As Object)
			Throw New NotSupportedException()
		End Sub

		Public Overrides Function ShouldSerializeValue(ByVal component As Object) As Boolean
			Return fDescriptor.ShouldSerializeValue(component)
		End Function
	End Class
End Namespace

File: CS/ReadOnlyCollectionWrapper/ReadOnlyCollectionWrapper.cs
using System;
using System.Collections;
using System.ComponentModel;
using System.Collections.Generic;

namespace DXSample {
    class ReadOnlyCollectionWrapper :Component, IBindingList, ITypedList {
        static readonly object fListChanged = new object();

        public ReadOnlyCollectionWrapper(IBindingList dataSource) {
            fDataSource = dataSource;
        }

        public ReadOnlyCollectionWrapper() { }

        IBindingList fDataSource;
        [AttributeProvider(typeof(IListSource))]
        public IBindingList DataSource {
            get { return fDataSource; }
            set {
                if (fDataSource == value) return;
                if (fDataSource != null)
                    fDataSource.ListChanged -= OnDataSourceListChanged;
                fDataSource = value;
                if (fDataSource != null)
                    fDataSource.ListChanged += OnDataSourceListChanged;
                RaiseListChanged(new ListChangedEventArgs(ListChangedType.Reset, -1));
            }
        }

        void RaiseListChanged(ListChangedEventArgs args) {
            ListChangedEventHandler handler = Events[fListChanged] as ListChangedEventHandler;
            if (handler != null)
                handler(this, args);
        }

        void OnDataSourceListChanged(object sender, ListChangedEventArgs e) {
            RaiseListChanged(e);
        }

        protected override void Dispose(bool disposing) {
            if (disposing)
                if (fDataSource != null) {
                    fDataSource.ListChanged -= OnDataSourceListChanged;
                    fDataSource = null;
                }
            base.Dispose(disposing);
        }

        #region IBindingList Members

        void IBindingList.AddIndex(PropertyDescriptor property) {
            fDataSource.AddIndex(property);
        }

        object IBindingList.AddNew() {
            throw new NotSupportedException();
        }

        bool IBindingList.AllowEdit {
            get { return true; }
        }

        bool IBindingList.AllowNew {
            get { return false; }
        }

        bool IBindingList.AllowRemove {
            get { return false; }
        }

        void IBindingList.ApplySort(PropertyDescriptor property, ListSortDirection direction) {
            fDataSource.ApplySort(property, direction);
        }

        int IBindingList.Find(PropertyDescriptor property, object key) {
            return fDataSource.Find(property, key);
        }

        bool IBindingList.IsSorted {
            get { return fDataSource.IsSorted; }
        }

        event ListChangedEventHandler IBindingList.ListChanged {
            add { Events.AddHandler(fListChanged, value); }
            remove { Events.RemoveHandler(fListChanged, value); }
        }

        void IBindingList.RemoveIndex(PropertyDescriptor property) {
            fDataSource.RemoveIndex(property);
        }

        void IBindingList.RemoveSort() {
            fDataSource.RemoveSort();
        }

        ListSortDirection IBindingList.SortDirection {
            get { return fDataSource.SortDirection; }
        }

        PropertyDescriptor IBindingList.SortProperty {
            get { return fDataSource.SortProperty; }
        }

        bool IBindingList.SupportsChangeNotification {
            get { return fDataSource.SupportsChangeNotification; }
        }

        bool IBindingList.SupportsSearching {
            get { return fDataSource.SupportsSearching; }
        }

        bool IBindingList.SupportsSorting {
            get { return fDataSource.SupportsSorting; }
        }

        #endregion

        #region IList Members

        int IList.Add(object value) {
            return fDataSource.Add(value);
        }

        void IList.Clear() {
            fDataSource.Clear();
        }

        bool IList.Contains(object value) {
            return fDataSource.Contains(value);
        }

        int IList.IndexOf(object value) {
            return fDataSource.IndexOf(value);
        }

        void IList.Insert(int index, object value) {
            fDataSource.Insert(index, value);
        }

        bool IList.IsFixedSize {
            get { return fDataSource.IsFixedSize; }
        }

        bool IList.IsReadOnly {
            get { return false; }
        }

        void IList.Remove(object value) {
            fDataSource.Remove(value);
        }

        void IList.RemoveAt(int index) {
            fDataSource.RemoveAt(index);
        }

        object IList.this[int index] {
            get { return fDataSource[index]; }
            set { fDataSource[index] = value; }
        }

        #endregion

        #region ICollection Members

        void ICollection.CopyTo(Array array, int index) {
            fDataSource.CopyTo(array, index);
        }

        int ICollection.Count {
            get { return fDataSource.Count; }
        }

        bool ICollection.IsSynchronized {
            get { return fDataSource.IsSynchronized; }
        }

        object ICollection.SyncRoot {
            get { return fDataSource.SyncRoot; }
        }

        #endregion

        #region IEnumerable Members

        IEnumerator IEnumerable.GetEnumerator() {
            return fDataSource.GetEnumerator();
        }

        #endregion

        #region ITypedList Members

        PropertyDescriptorCollection ITypedList.GetItemProperties(PropertyDescriptor[] listAccessors) {
            ITypedList tList = fDataSource as ITypedList;
            PropertyDescriptorCollection properties;
            if (tList == null)
                if (fDataSource.Count == 0 || listAccessors != null) properties = new PropertyDescriptorCollection(new PropertyDescriptor[0]);
                else properties = TypeDescriptor.GetProperties(fDataSource[0]);
            else
                properties = ((ITypedList)fDataSource).GetItemProperties(listAccessors);
            List<PropertyDescriptor> result = new List<PropertyDescriptor>();
            foreach (PropertyDescriptor property in properties)
                result.Add(new ReadOnlyPropertyDescriptor(property));
            return new PropertyDescriptorCollection(result.ToArray());
        }

        string ITypedList.GetListName(PropertyDescriptor[] listAccessors) {
            ITypedList tList = fDataSource as ITypedList;
            if (tList == null)
                if (fDataSource.Count == 0 || listAccessors != null) return string.Empty;
                else return fDataSource[0].GetType().Name;
            return ((ITypedList)fDataSource).GetListName(listAccessors);
        }

        #endregion
    }

    public class ReadOnlyPropertyDescriptor :PropertyDescriptor {
        private PropertyDescriptor fDescriptor;

        public ReadOnlyPropertyDescriptor(PropertyDescriptor descr) : base(descr) {
            fDescriptor = descr;
        }

        public override bool CanResetValue(object component) {
            return false;
        }

        public override Type ComponentType {
            get { return fDescriptor.ComponentType; }
        }

        public override object GetValue(object component) {
            return fDescriptor.GetValue(component);
        }

        public override bool IsReadOnly {
            get { return true; }
        }

        public override Type PropertyType {
            get { return fDescriptor.PropertyType; }
        }

        public override void ResetValue(object component) {
            throw new NotSupportedException();
        }

        public override void SetValue(object component, object value) {
            throw new NotSupportedException();
        }

        public override bool ShouldSerializeValue(object component) {
            return fDescriptor.ShouldSerializeValue(component);
        }
    }
}
